File: packages/ruby-sdk/lib/opensend/resources/templates.rb
# frozen_string_literal: true

module OpenSend
  module Resources
    class Templates
      def initialize(http)
        @http = http
      end

      # POST /templates
      def create(params)
        @http.post("/templates", normalize(params))
      end

      # GET /templates
      def list(limit: nil, after: nil, search: nil, status: nil)
        params = {}
        params[:limit] = limit unless limit.nil?
        params[:after] = after if after
        params[:search] = search if search
        params[:status] = status if status
        @http.get("/templates", params)
      end

      # GET /templates/:id_or_alias
      def get(id_or_alias)
        @http.get("/templates/#{id_or_alias}")
      end

      # PATCH /templates/:id_or_alias
      def update(id_or_alias, params)
        @http.patch("/templates/#{id_or_alias}", normalize(params))
      end

      # DELETE /templates/:id_or_alias
      def delete(id_or_alias)
        @http.delete("/templates/#{id_or_alias}")
      end

      # POST /templates/:id_or_alias/publish
      def publish(id_or_alias)
        @http.post("/templates/#{id_or_alias}/publish")
      end

      # POST /templates/:id_or_alias/duplicate
      def duplicate(id_or_alias)
        @http.post("/templates/#{id_or_alias}/duplicate")
      end

      private

      ALIASES = {
        "replyTo" => "reply_to",
        "previewText" => "preview_text",
        "fallbackValue" => "fallback_value"
      }.freeze

      def normalize(params)
        return {} if params.nil? || params.empty?

        result = {}
        params.each do |k, v|
          key = k.to_s
          result[ALIASES.fetch(key, key)] = v
        end
        result
      end
    end
  end
end

File: packages/python-sdk/src/opensend/templates.py
"""Templates resource for the OpenSend Python SDK."""

from __future__ import annotations

from typing import Any, Mapping, Optional, cast

from ._http import HttpClient, JsonObject
from ._types import (
    CreateTemplatePayload,
    CreateTemplateResponse,
    DeleteTemplateResponse,
    DuplicateTemplateResponse,
    PublishTemplateResponse,
    TemplateListOptions,
    TemplateListResponse,
    TemplateResponse,
    UpdateTemplatePayload,
    UpdateTemplateResponse,
)


def _normalize_template_payload(payload: Mapping[str, Any]) -> JsonObject:
    """Resolve from_email → from for the API."""
    result = dict(payload)
    from_email = result.pop("from_email", None)
    if "from" not in result and from_email is not None:
        result["from"] = from_email
    return result


class TemplatesResource:
    """CRUD + publish/duplicate for the /templates namespace."""

    def __init__(self, client: HttpClient) -> None:
        self._client = client

    def create(self, payload: CreateTemplatePayload) -> CreateTemplateResponse:
        """Create a new email template."""
        return cast(
            CreateTemplateResponse,
            self._client.request("POST", "/templates", _normalize_template_payload(payload)),
        )

    def list(self, options: Optional[TemplateListOptions] = None) -> TemplateListResponse:
        """List templates with optional filters and pagination."""
        opts = options or {}
        query: dict[str, str] = {}
        if opts.get("limit") is not None:
            query["limit"] = str(opts["limit"])
        if opts.get("after"):
            query["after"] = opts["after"]  # type: ignore[assignment]
        if opts.get("search"):
            query["search"] = opts["search"]  # type: ignore[assignment]
        if opts.get("status"):
            query["status"] = opts["status"]  # type: ignore[assignment]
        return cast(
            TemplateListResponse,
            self._client.request("GET", "/templates", params=query or None),
        )

    def get(self, id_or_alias: str) -> TemplateResponse:
        """Retrieve a template by ID or alias."""
        return cast(
            TemplateResponse,
            self._client.request("GET", f"/templates/{id_or_alias}"),
        )

    def update(
        self, id_or_alias: str, payload: UpdateTemplatePayload
    ) -> UpdateTemplateResponse:
        """Update a template's content or metadata."""
        return cast(
            UpdateTemplateResponse,
            self._client.request(
                "PATCH",
                f"/templates/{id_or_alias}",
                _normalize_template_payload(payload),
            ),
        )

    def delete(self, id_or_alias: str) -> DeleteTemplateResponse:
        """Delete a template by ID or alias."""
        return cast(
            DeleteTemplateResponse,
            self._client.request("DELETE", f"/templates/{id_or_alias}"),
        )

    def publish(self, id_or_alias: str) -> PublishTemplateResponse:
        """Publish the current draft version of a template."""
        return cast(
            PublishTemplateResponse,
            self._client.request("POST", f"/templates/{id_or_alias}/publish"),
        )

    def duplicate(self, id_or_alias: str) -> DuplicateTemplateResponse:
        """Duplicate a template."""
        return cast(
            DuplicateTemplateResponse,
            self._client.request("POST", f"/templates/{id_or_alias}/duplicate"),
        )
